Stdlib documentation

Mte90 opened the pull request:
Ref: https://github.com/amber-lang/amber-docs/issues/38

I need a bit of help to improve the documentation also because I didn't understood the purpose of some function.

I think that @CymDeveloppement can help me about it as some of them are by him.

Mte90 commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
-    let result = [Text]
-    unsafe $IFS=\$'\n' read -rd '' -a {nameof result} <<<"\${nameof text}"$
-    return result
+    return split(text, "\n")
```
I improved that one as I see it was the same

Mte90 commented on src/std/fs.ab:
```diff
@@ -1,55 +1,66 @@
     }
 
-    echo "The file {path} doesn't exist!"
+    error("The file {path} doesn't exist!")
```
We can use the new function to color the output

CymDeveloppement replied:
also just fail

Mte90 commented on src/std/fs.ab:
```diff
@@ -1,55 +1,66 @@
     }
 
-    echo "The file {origin} doesn't exist!"
+    error("The file {origin} doesn't exist!")
```
Wondering if we want to `exit` the program when it fails, after all the program is crashing otherwise.

CymDeveloppement replied:
I think just this function can be fail.
if you exit user lost control and cannot catch error

Mte90 replied:
We can show the warning and exit, like most of the scripting languages does when a file doens't exist

CymDeveloppement replied:
if we display warnings in functions, this should be able to be deactivated by user.

Mte90 replied:
It is something we need to decide in another ticket.

Good point

CymDeveloppement replied:
```
pub fun create_symbolic_link(path: Text): Bool {
  silent $ln -s "{origin}" "{destination}"$ failed {
          return false
  }    
  return true
}
```

Mte90 commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
 
-
+/// Split the string and return an array
 pub fun split(text: Text, delimiter: Text): [Text] {
```
I am not sure about the array as return as I see `Text`

Mte90 commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
 }
 
+/// Split by space separator
 pub fun words(text: Text): [Text] {
```
It should return an array instead of a string?

CymDeveloppement replied:
an array of all words

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -15,59 +16,71 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
 }
 
+/// ?
 pub fun shell_isset(name: Text): Bool {
```
return true if variable is set also return false

KrosFire replied:
Checks if a specified environment variable is set and is not empty.

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     return false
 }
 
+/// ?
```
printf c command with format in first parameters and array of string
https://www.man7.org/linux/man-pages/man1/printf.1.html

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     unsafe $printf "\$\{{nameof args}[@]}"$
 }
 
+/// ?
```
escape special character for use in printf format.
I will give more explication

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     printf("\x1b[1;3;97;43m %s \x1b[0m\n", [message])
 }
 
+/// Print a string as Error
```
and exit if no exit code is provided or exit code greater than 0

KrosFire commented on src/std/array.ab:
```diff
@@ -1,3 +1,4 @@
+/// Get the index of the first value found in the array specified
```
```suggestion
/// Returns index of the first value found in the specified array
/// If the value is not found, the function returns -1
```

KrosFire commented on src/std/array.ab:
```diff
@@ -7,6 +8,7 @@ pub fun array_first_index(array, value): Num {
     return -1
 }
 
+/// Get the total of the values found in the array
```
```suggestion
/// Searches for all occurrences of a given value in an array and returns their indices.
```

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     return unsafe $echo \${nameof text} | sed -e 's/\\\\/\\\\\\\\/g' -e "s/%/%%/g"$
 }
 
+/// Print a string with formatting options
 pub fun text_shell(message: Text, style: Num, fg: Num, bg: Num): Text {
     return "\x1b[{style};{fg};{bg}m{printf_escape(message)}\x1b[0m"
 }
```
not printing but "prepare" this function return a string can be passed to printf function.
`message` must be escaped before send to printf function
if you want to print `\version` in bold `\` must be escaped

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     return "\x1b[{style};{fg};{bg}m{printf_escape(message)}\x1b[0m"
 }
 
+/// Print a string as bold
 pub fun text_bold(message: Text): Text {
     return "\x1b[1m{printf_escape(message)}\x1b[0m"
 }
```
not printing

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     return "\x1b[1m{printf_escape(message)}\x1b[0m"
 }
 
+/// Print a string as italic
 pub fun text_italic(message: Text): Text {
     return "\x1b[3m{printf_escape(message)}\x1b[0m"
 }
```
not printing

CymDeveloppement commented on src/std/env.ab:
```diff
@@ -76,52 +89,59 @@ pub fun is_root(): Bool {
     return "\x1b[3m{printf_escape(message)}\x1b[0m"
 }
 
+/// Print a string as underlined
 pub fun text_underlined(message: Text): Text {
     return "\x1b[4m{printf_escape(message)}\x1b[0m"
 }
```
not printing

KrosFire commented on src/std/array.ab:
```diff
@@ -17,11 +19,13 @@ pub fun array_search(array, value): [Num] {
     return result
 }
 
+/// Check if the value is in the array
 pub fun in_array(array, value): Bool {
     let result = array_first_index(array, value)
     return result >= 0
 }
 
+/// Check if the value is in the array
```
Why do we have both of those functions?

Mte90 replied:
Do you mean `includes` and `in_array`?

To me they do the same things and probably initially there was one and later the other one it was added.
I guess that we can remove `includes` as the other one is better.

Ph0enixKM replied:
I prefer the name `includes` over `in_array` as it's more natural

KrosFire commented on src/std/env.ab:
```diff
@@ -15,59 +16,71 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
     return ""
 }
```
There is a bug here on line 11. It should be:
```
if file_exist(file) {
```

Mte90 replied:
good catch!

KrosFire commented on src/std/env.ab:
```diff
@@ -1,6 +1,7 @@
 import * from "std/fs"
 import * from "std/text"
 
+/// Get the variable from the environment, otherwise check from the input file
```
otherwise?

```suggestion
/// Retrieves the value of an environment variable, optionally sourcing it from a file if not already set.
```

KrosFire commented on src/std/env.ab:
```diff
@@ -15,59 +16,71 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
     return $echo \$\{!{nameof name}}$?
 }
 
+/// Set a variable?
 pub fun shell_var_set(name: Text, val: Text): Null {
     $export \${nameof name}="\${nameof val}" 2> /dev/null$?
 }
 
+/// Get a variable?
 pub fun shell_var_get(name: Text): Text {
     return $echo \$\{!{nameof name}}$?
 }
 
+/// Remove a variable?
```
Sounds right to me

KrosFire commented on src/std/env.ab:
```diff
@@ -15,59 +16,71 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
     return "\$REPLY"
 }
 
+/// Check if the command passed fails
```
```suggestion
/// Checks if the command has failed
```

KrosFire commented on src/std/fs.ab:
```diff
@@ -1,55 +1,66 @@
     return $< "{path}"$?
 }
 
+/// Write the content to the file, doesn't check if the file exist
```
second part in new line maybe?

KrosFire commented on src/std/fs.ab:
```diff
@@ -1,55 +1,66 @@
     return false
 }
 
+/// Create a directory with the full path
```
```suggestion
/// Create a directory with all intermediate directories as required
```

KrosFire commented on src/std/fs.ab:
```diff
@@ -1,55 +1,66 @@
     return false
 }
 
+/// Change the owner of the file, if the file doesn't exist return a boolean
```
```suggestion
/// Change the owner of the file, if the file doesn't exist return false
```

KrosFire commented on src/std/http.ab:
```diff
@@ -1,5 +1,6 @@
 import * from "std/env"
 
+/// It will download the file using (if present) in the order: `curl`, `wget`, `aria2c`
```
```suggestion
/// Downloads a file from a given URL and saves it to a specified path using available command-line tools.
///
/// This function attempts to download a file from the provided URL and save it to the specified path.
/// It checks for the availability of common command-line tools (`curl`, `wget`, and `aria2c`) and uses the first available tool to perform the download.
/// If none of the tools are available, the function returns `false`.
```

Mte90 commented on src/std/env.ab:
```diff
@@ -15,59 +16,71 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
     return $echo \$\{!{nameof name}}$?
 }
 
+/// Set a variable?
```
@CymDeveloppement

CymDeveloppement replied:
sorry, yes this function set a shell variable

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
+/// Replace the content just once
```
```suggestion
/// Finds the first occurrence of a pettern in the content and replaces it with provided replacement text
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return unsafe $echo "\$\{source/{pattern}/{replacement}}"$
 }
 
+// Replace the content multiple times
```
```suggestion
/// Replaces all occurences of a pattern in the content content with provided replacement text
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return unsafe $echo "\$\{source//{pattern}/{replacement}}"$
 }
 
+/// Replace with regex using `sed`
```
```suggestion
/// Replaces all occurences of a regex pattern in the content content with provided replacement text
///
/// Function uses `sed`
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
 }
 
-
+/// Split the string and return an array
```
```suggestion
/// This function splits the input `text` into an array of substrings using the specified `delimiter`. 
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return result
 }
 
+/// Split a string with newlines
```
```suggestion
/// Splits a `text` into an array of substrings based on newline characters.
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return trim_left(trim_right(text))
 }
 
+/// Lowercase the string input using `tr`
```
I think we should be saying "text" instead of "string" in Amber

Mte90 replied:
The problem is that `string` we are talking about a type, `text` is generic for that context.
Sure in Amber has that meaning, so yes we need to decide.

Ph0enixKM replied:
Let's just stick to text everywhere

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return unsafe $echo "{text}" | tr '[:lower:]' '[:upper:]'$
 }
 
+/// ?
```
```suggestion
/// Attempts to parse a given text into a number, returning the parsed number or zero if parsing fails.
```

KrosFire commented on src/std/text.ab:
```diff
@@ -1,62 +1,73 @@
     return text as Num
 }
 
+/// ?
```
```suggestion
/// Splits a text into an array of individual characters.
```

KrosFire commented on src/std/env.ab:
```diff
@@ -16,12 +16,12 @@ pub fun get_env_var(var: Text, file: Text = ".env"): Text {
 }
 
-/// ?
+/// If the variable is set in Bash return as boolean
```
Checks if a specified environment variable is set and is not empty.

Mte90 commented:
Thanks @CymDeveloppement and @KrosFire for the help :-)

We need just to see a couple of points but I think we just need a review from @Ph0enixKM so we can finish that PR :-)

CymDeveloppement commented on src/std/fs.ab:
```diff
@@ -1,55 +1,71 @@
     }
 
-    echo "The file {path} doesn't exist!"
+    error("The file {path} doesn't exist!")
```
I think this is not necessary if the function return false.
maybe is not an error if is an user input.
```
pub fun make_executable(path: Text): Bool {
  silent $chmod +x "{path}"$ failed {
          return false
  }    
  return true
}
```

CymDeveloppement replied:
also this "resolve" circular import problem

Mte90 replied:
I don't like silent as we can't be sure if it is executed and usually a check avoid a bash error in this case generated by chmod.
I think that is part of the discussion we have in the dedicated discussion about the error reporting.

About the circular I opened a ticket as it shouldn't happen as I am just importing libraries.

CymDeveloppement replied:
maybe for this PR we can keep the function the same. this introduced a breaking change or the function call must be : `error("The file {path} doesn't exist!", 0)`.
I think we can open discussion about error displaying and verbose level management.

Ph0enixKM reviewed (changes requested):
Okay I see now. Let's wait for the PR first

Mte90 commented:
So it is ready to get approved :-)

b1ek commented:
what about the #360? i dont think that has been addressed yet

Mte90 commented:
Now the PR is free from that issue as I removed the code (the CI works).
In this way we can merge it and do the new release as there is no blockers.
We implemented too many things in this alpha.
When that will be implemented we can improve the stdlib code reusing stuff in the various files.

Ph0enixKM commented:
I'm still not a big fan of having the `len` function in a `std/text` module and nothing like that in `std/array` module as it works for both text and arrays. How about creating `std/core` that has the `len` function and perhaps some more of the most core and atomic functions? @Mte90 @b1ek @KrosFire

Mte90 commented:
I think that can create confusion.
Usually is a native function in a script language, I think that we can keep as it is now and later move as builtin.

KrosFire commented:
> I'm still not a big fan of having the `len` function in a `std/text` module and nothing like that in `std/array` module as it works for both text and arrays. How about creating `std/core` that has the `len` function and perhaps some more of the most core and atomic functions? @Mte90 @b1ek @KrosFire

Let's move it to separate issue. I think that `std/core` or something like that is a good idea

Ph0enixKM reviewed (approved):
Looking good
